lib/ai/model/tokencount.go

```diff
@@ -0,0 +1,273 @@
+// SynchronousTokenCounter count tokens synchronously, while
+// AsynchornousTokenCounter supports the streaming use-cases.
+type TokenCount struct {
+	Prompt     TokenCounters
+	Completion TokenCounters
+}
+
+// AddPromptCounter adds a TokenCounter to the Prompt list.
+func (tc *TokenCount) AddPromptCounter(prompt TokenCounter) {
+	if prompt != nil {
+		tc.Prompt = append(tc.Prompt, prompt)
+	}
+}
+
+// AddCompletionCounter adds a TokenCounter to the Completion list.
+func (tc *TokenCount) AddCompletionCounter(completion TokenCounter) {
+	if completion != nil {
+		tc.Completion = append(tc.Completion, completion)
+	}
+}
+
+// CountAll iterates over all counters and returns how many prompt and
+// completion tokens were used. As completion token counting can require waiting
+// for a response to be streamed, the caller should pass a context and use it to
+// implement some kind of deadline to avoid hanging infinitely if something goes
+// wrong (e.g. use `context.WithTimeout()`).
+func (tc *TokenCount) CountAll(ctx context.Context) (int, int, error) {
+	prompt, err := tc.Prompt.CountAll(ctx)
+	if err != nil {
+		return 0, 0, trace.Wrap(err)
+	}
+	completion, err := tc.Completion.CountAll(ctx)
+	if err != nil {
+		return 0, 0, trace.Wrap(err)
+	}
+	return prompt, completion, nil
+}
+
+// NewTokenCount initializes a new TokenCount struct.
+func NewTokenCount() *TokenCount {
+	return &TokenCount{
+		Prompt:     TokenCounters{},
+		Completion: TokenCounters{},
+	}
+}
+
+// TokenCounter is an interface for all token counters, regardless of the kind
+// of token they count (prompt/completion) or the tokenizer used.
+// TokenCount must be idempotent.
+type TokenCounter interface {
+	TokenCount(ctx context.Context) (int, error)
+}
+
+// TokenCounters is a list of TokenCounter and offers function to iterate over
+// all counters and compute the total.
+type TokenCounters []TokenCounter
+
+// CountAll iterates over a list of TokenCounter and returns the sum of the
+// results of all counters. As the counting process might be blocking/take some
+// time, the caller should set a Deadline on the context.
+func (tc TokenCounters) CountAll(ctx context.Context) (int, error) {
+	var total int
+	for _, counter := range tc {
+		count, err := counter.TokenCount(ctx)
+		if err != nil {
+			return 0, trace.Wrap(err)
+		}
+		total += count
+	}
+	return total, nil
+}
+
+// PromptTokenCounter implements the TokenCounter interface and counts tokens
+// for a given prompt. Tokenization happens on initialization.
+type PromptTokenCounter struct {
+	count int
+}
+
+// TokenCount implements the TokenCounter interface.
+func (tc *PromptTokenCounter) TokenCount(_ context.Context) (int, error) {
+	return tc.count, nil
+}
+
+// NewPromptTokenCounter takes a list of openai.ChatCompletionMessage and
+// computes how many tokens are used by sending those messages to the model.
+func NewPromptTokenCounter(prompt []openai.ChatCompletionMessage) (*PromptTokenCounter, error) {
```

Comment:
do we need to split into two classes here and add a relatively large amount of boilerplate compared to the business logic here? is it sufficient to maintain one token counter class that keeps track of both? It's just a lot of newlines, godocs and various misc methods that don't really do anything.

Reply:
I merged prompt and synchronous token counters in the same type and eliminated the struct.